dotnet/src/InternalUtilities/connectors/AI/FunctionCallsProcessor.cs

```diff
@@ -0,0 +1,342 @@
+/// Class responsible for providing function calling configuration and processing AI function calls. As part of the processing, it will:
+/// 1. Iterate over <see cref="FunctionCallContent"/> items representing AI model function calls in the <see cref="ChatMessageContent.Items"/> collection.
+/// 2. Look up each function in the <see cref="Kernel"/>.
+/// 3. Invoke the auto function invocation filter, if registered, for each function.
+/// 4. Invoke each function and add the function result to the <see cref="ChatHistory"/>.
+/// </summary>
+[ExcludeFromCodeCoverage]
+internal sealed class FunctionCallsProcessor
+{
+    /// <summary>
+    /// The maximum number of auto-invokes that can be in-flight at any given time as part of the current
+    /// asynchronous chain of execution.
+    /// </summary>
+    /// <remarks>
+    /// This is a fail-safe mechanism. If someone accidentally manages to set up execution settings in such a way that
+    /// auto-invocation is invoked recursively, and in particular where a prompt function is able to auto-invoke itself,
+    /// we could end up in an infinite loop. This const is a backstop against that happening. We should never come close
+    /// to this limit, but if we do, auto-invoke will be disabled for the current flow in order to prevent runaway execution.
+    /// With the current setup, the way this could possibly happen is if a prompt function is configured with built-in
+    /// execution settings that opt-in to auto-invocation of everything in the kernel, in which case the invocation of that
+    /// prompt function could advertize itself as a candidate for auto-invocation. We don't want to outright block that,
+    /// if that's something a developer has asked to do (e.g. it might be invoked with different arguments than its parent
+    /// was invoked with), but we do want to limit it. This limit is arbitrary and can be tweaked in the future and/or made
+    /// configurable should need arise.
+    /// </remarks>
+    private const int MaxInflightAutoInvokes = 128;
+
+    /// <summary>
+    /// The maximum number of function auto-invokes that can be made in a single user request.
+    /// </summary>
+    /// <remarks>
+    /// After this number of iterations as part of a single user request is reached, auto-invocation
+    /// will be disabled. This is a safeguard against possible runaway execution if the model routinely re-requests
+    /// the same function over and over.
+    /// </remarks>
+    private const int MaximumAutoInvokeAttempts = 128;
+
+    /// <summary>Tracking <see cref="AsyncLocal{Int32}"/> for <see cref="MaxInflightAutoInvokes"/>.</summary>
+    /// <remarks>
+    /// It is temporarily made internal to allow code that uses the old function model to read it and decide whether to continue auto-invocation or not.
+    /// It should be made private when the old model is deprecated.
+    /// </remarks>
+    internal static readonly AsyncLocal<int> s_inflightAutoInvokes = new();
```

Comment:
Given this is static what happens when I have a high volume application with potentially 000's of requests using function calling be executed in parallel?

Reply:
Despite the field is static its value is unique per execution flow, so it 000's of request hit it in parallel each request will see it's unique value.